### PR TITLE
Improvements for Edge cases

### PR DESCRIPTION
**Pull request by Stono**
This PR contains numerous fixes for things I've encountered whilst running against an ArcSight Connector receiving Syslog entries, such as some fields not always being populated, or white space messing up parsing.

I've added tests and fixes for those cases as they arose.

I could not for the life of me get jruby working on my host, so as a result have also created a JRuby development environment docker file and added instructions on how to use it in the README.md file.

Would be great if you could merge this in and release to rubygems as at the moment I'm having to use my published fork, plus i'm sure other people would want these fixes.

### CODE REVIEW

**Comment by Stono** [comment]
Please note that this now also includes some pretty big changes so I appreciate if these change aren't merged in. 

I decided that the extension dictionary should be under a hash of its own called 'extension' rather than a load of fields of cef_ext_thing.

I've also removed the requirement to specify if we're expecting syslog headers, and just detecting them instead, simplifying the use in my opinion

**Comment by jsvd** [comment]
@Stono wow so many improvements, sorry for overlooking this for so long.
I have some suggestions to make on your PR, let's iterate on it a bit then merge :)

File: lib/logstash/codecs/cef.rb [hunk 1]
```diff
@@ -1,89 +1,86 @@
...
+    config_name "cef"
+    config :signature, :validate => :string, :default => "Logstash"
+    config :name, :validate => :string, :default => "Logstash"
+    config :sev, :validate => :number, :default => 6
+
+    config :fields, :validate => :array
+
+    public
+    def initialize(params={})
+        super(params)
```
Comment (jsvd) [inline lib/logstash/codecs/cef.rb]: just for consistency sake on all other logstash and plugins codebase, can you reformat as 2 space based indentation?

File: lib/logstash/codecs/cef.rb [hunk 2]
```diff
@@ -1,89 +1,86 @@
...
-      return name + "=" + val.to_json
-    else
-      return name + "=" + val
+
+    public
+    def decode(data)
+        # Strip any quotations at the start and end, flex connectors seem to send this
+        if data[0] == "\""
+            data = data[0..-2]
+            data.slice!(0)
```
Comment (jsvd) [inline lib/logstash/codecs/cef.rb]: Doing  `data = data[1..-2]` should be equivalent to these 2 operations

File: lib/logstash/codecs/cef.rb [hunk 3]
```diff
@@ -1,89 +1,86 @@
...
+        end
+        event = LogStash::Event.new()
```
Comment (jsvd) [inline lib/logstash/codecs/cef.rb]: for style consistency, no need to have parenthesis here

File: lib/logstash/codecs/cef.rb [hunk 4]
```diff
@@ -1,89 +1,86 @@
...
+        # Try and parse out the syslog header if there is one
+        if event['cef_version'].include? ' '
+            event['syslog'], event['cef_version'] = event['cef_version'].split(' ')
+        end
+
+        # Get rid of the CEF bit in the version
+        event['cef_version'].sub! /^CEF:/, ''
+
+        # Strip any whitespace from the message
+        message = message.to_s.strip
```
Comment (jsvd) [inline lib/logstash/codecs/cef.rb]: is the .to_s necessary?

File: lib/logstash/codecs/cef.rb [hunk 5]
```diff
@@ -1,89 +1,86 @@
...
+
+        # If the last KVP has no value, add an empty string, this prevents hash errors below
+        if message[-1, 1] == "="
```
Comment (jsvd) [inline lib/logstash/codecs/cef.rb]: suggestion:  
`message[-1] == "="` or `message.end_with?("=")`

File: lib/logstash/codecs/cef.rb [hunk 6]
```diff
@@ -1,89 +1,86 @@
...
+            message=message + ' '
+        end
+
+        # Now parse the key value pairs into it
+        extensions = {}
+        if message.length != 0 and message.include? "="
+            message = message.split(/ ([\w\.]+)=/)
```
Comment (jsvd) [inline lib/logstash/codecs/cef.rb]: My suggestion for breaking the string into a Hash:

```ruby
irb(main):024:0> message = "src=10.0.0.192 dst=12.121.122.82 spt=1232"
=> "src=10.0.0.192 dst=12.121.122.82 spt=1232"
irb(main):025:0> extensions = Hash[*(message.split(/ |=/))]
=> {"src"=>"10.0.0.192", "dst"=>"12.121.122.82", "spt"=>"1232"}
```

**Comment by Stono** [comment]
Hi @jsvd 
I've made the changes you suggested, could tell that I was rushing when I wrote this originally and that ruby is by no means my primary language haha!

The only thing I didn't implement was your suggestion on the regex for the extention dictionary, the changes resulted in a failed test (yey for tests) where the message contained an escaped |, not being used as a delimiter, the current implementation works and all the tests pass so I decided to stick with that rather than trying to debug the regex you suggested.

```
[root@devenv-jruby master] /storage
# jruby -S bundle exec rspec
Using Accessor#strict_set for specs
Run options: exclude {:redis=>true, :socket=>true, :performance=>true, :couchdb=>true, :elasticsearch=>true, :elasticsearch_secure=>true, :export_cypher=>true, :integration=>true, :windows=>true}
.....F*

Pending:
  LogStash::Codecs::CEF#encode should assert all header fields are present
    # Not yet implemented
    # ./spec/codecs/cef_spec.rb:13

Failures:

  1) LogStash::Codecs::CEF#decode should be OK with escaped pipes in the message
     Failure/Error: subject.decode(escaped_pipes) do |e|
     ArgumentError:
       odd number of arguments for Hash
     # ./lib/logstash/codecs/cef.rb:46:in `decode'
     # ./spec/codecs/cef_spec.rb:64:in `(root)'

Finished in 0.061 seconds
7 examples, 1 failure, 1 pending

Failed examples:

rspec ./spec/codecs/cef_spec.rb:63 # LogStash::Codecs::CEF#decode should be OK with escaped pipes in the message
```

File: lib/logstash/codecs/cef.rb [hunk 7]
```diff
@@ -20,31 +15,45 @@ def initialize(params={})
...
+      extensions[key] = value
 
-      key, value = message.shift.split('=',2)
-      @logger.debug(message)
-      kv = Hash[*message]
-      @logger.debug(kv)
-      addKey(kv,key,value)
-      event.to_hash.merge!(Hash[kv.map{ |k,v| ["cef_ext_"+k,v] }])
-    end #
+      Hash[*message].each{|k, v|
```
Comment (jsvd) [inline lib/logstash/codecs/cef.rb]: just for style consistency, can you either switch to 1 line {} or block do;end such as?

```ruby
Hash[*message].each{|k, v| extensions[k] = v }
```

or 

```ruby
Hash[*message].each do |k, v|
  extensions[k] = v
end
```

**Comment by jsvd** [comment]
@Stono looks great! just a minor comment and I'll merge

**Comment by Stono** [comment]
@jsvd Done and Done.  Take it you'll version bump and push to rubygems?

**Comment by Stono** [comment]
Also, any objection to adding me to the s.authors?

File: lib/logstash/codecs/cef.rb (hunk 4 above)
Comment (Stono) [inline lib/logstash/codecs/cef.rb]: I forgot to mention, I didn't implement this as if there was no "message" passed back in the CEF event (therefore no extension dictionary) then message would be nil which would result an error with the .strip, .to_s changes the nil to "" which then ensures the code after works out.

Could probably switch this out for a if statement asserting on nil I suppose, however I'm not sure what would be more performant on messages which do contain extra data (almost all), a `.to_s` or an `if not nil?`

**Comment by Stono** [comment]
@jsvd hold off on the merge for two mins... ;)

**Comment by jsvd** [comment]
@Stono the project has [a CONTRIBUTORS](https://github.com/logstash-plugins/logstash-codec-cef/blob/master/CONTRIBUTORS) file for this very purpose, feel free to add yourself to the list in this PR :)

**Comment by Stono** [comment]
@jsvd all done, I wanted to do something about your earlier to_s comment, i've reduced the cyclomatic complexity a little by restructuring the code slightly which also resulted in me being able to remove the to_s.  I've also added another test case which covers what the "to_s" was protecting against - a nill message (extension dictionary).

Think we're good to go here :+1:

**Comment by jsvd** [comment]
Can you rebase against master?

**Comment by jsvd** [comment]
@Stono and, after that, as a last step I need you to do [step 2 of the contributing guide](https://github.com/elastic/logstash/blob/master/CONTRIBUTING.md#contribution-steps).

However I see that this PR has commits from 2 different emails, so can I ask you to force push into this branch these commits with only one of the emails?

Here is one way to do it (after rebasing pr branch against master):
```
[in pr branch] $ git format-patch --stdout master > my.patch  # create a patch file with all changes from this pr
               # edit the my.patch file to make sure all commits have the same email
[in pr branch] $ git checkout origin/master -b tmp_pr # create a branch off of the origin/master and switch to it
[in tmp_pr   ] $ cat my.patch | git am   # apply git commits from the patch file
[in tmp_pr   ] $ git push -f Hewlett-Packard-ESS master  # overwrite original PR
```

**Comment by Stono** [comment]
@jsvd Is that really required?  They're both valid email addresses, one represents stuff I've done during office/work hours and the other is stuff I've done personally.

**Comment by jsvd** [comment]
@Stono or..can you instead add the corporate email account to your github account so we can match them together?

**Comment by ryanbreed** [comment]
yo @jsvd @Stono - where's this PR at? these commits are pretty useful.

**Comment by Stono** [comment]
@jsvd I've added my corporate email to my account so you should be good to merge now.

cc. @ryanbreed

**Comment by jsvd** [comment]
merged by abc300f7d91e367f8c67f13863d9f8a34057f9b8
thanks for the commits and patience @Stono

**Comment by Stono** [comment]
@jsvd no worries

**Comment by Stono** [comment]
@jsvd  any chance of pushing it to rubygems/version bump? https://rubygems.org/gems/logstash-codec-cef/versions/0.1.3

**Comment by jsvd** [comment]
Done https://rubygems.org/gems/logstash-codec-cef/versions/0.1.5

**Comment by Stono** [comment]
Thanks man :)

**Pull request closed without merging**